File: ros_ws/src/helpers/utcoupe_metrics/src_py/utcoupe_metrics/metric_types/counter.py
#!/usr/bin/env python
# -*-coding:Utf-8 -*

__author__ = "Gaëtan Blond"
__date__ = "20/03/20"

import rospy

from std_msgs.msg import Float64

class Counter(object):
    def __init__(self, topic_name, start_value = 0.0):
        self.m_publisher = rospy.Publisher(topic_name, Float64, queue_size=10)
        self.m_value = start_value

    def inc(self, val_inc = 1.0):
        if val_inc < 0.0:
            raise Exception("counter increment value must be positive")

        self.m_value += val_inc

    def publish_value(self):
        self.m_publisher.publish(self.m_value)

File: ros_ws/src/helpers/utcoupe_metrics/src/utcoupe_metrics/metric_types/counter.cpp
#include "utcoupe_metrics/metric_types/counter.h"

#include <std_msgs/Float64.h>

#include <exception>

using namespace utcoupe_metrics;

metric_types::Counter::Counter(ros::NodeHandle & nh, const std::string& topic_name, double start_value):
    m_value(start_value) {
    m_publisher = nh.advertise<std_msgs::Float64>(topic_name, 10);
}

void utcoupe_metrics::metric_types::Counter::inc(double val_inc)
{
    if (val_inc < 0) {
        throw std::runtime_error("a counter cannot decrease its value");
    }
    m_value += val_inc;
}

void utcoupe_metrics::metric_types::Counter::publish()
{
    std_msgs::Float64 msg;
    msg.data = m_value;
    m_publisher.publish(msg);
}
